File: leetcode/Python/furthest-building-you-can-reach.py
from queue import PriorityQueue
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        que = PriorityQueue()
        cur,t = heights[0],0
        for i in range(1,len(heights)):
            t = heights[i]-cur
            if t > 0:
                que.put(-t)
                bricks-=t
                while bricks<0 and ladders>0:
                    bricks-=que.get()
                    ladders-=1
                if bricks < 0:
                    return i-1
            cur = heights[i]
        return len(heights)-1

File: leetcode/C++/furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int b, int l) {
        priority_queue<int> pq;
        int cur = heights[0], t;
        for(int i=1; i<heights.size(); cur=heights[i++]){
            t=heights[i]-cur;
            if(t>0) {
                pq.push(t);
                b-=t;
                for(; b<0 && l>0; --l,pq.pop()){b+=pq.top();}
                if(b<0) return i-1;
            }
        }
        return heights.size()-1;
    }
};